SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc

```diff
@@ -0,0 +1,334 @@
+    algomap_.clear();
+  }
+  Phase2TrackerDigitizer::~Phase2TrackerDigitizer() {  
+    edm::LogInfo("Phase2TrackerDigitizer") << "Destroying the Digitizer";
+  }
+  void
+  Phase2TrackerDigitizer::accumulatePixelHits(edm::Handle<std::vector<PSimHit> > hSimHits,
+				       size_t globalSimHitIndex,const unsigned int tofBin) {
+    if (hSimHits.isValid()) {
+      std::set<unsigned int> detIds;
+      std::vector<PSimHit> const& simHits = *hSimHits.product();
+      for (auto it = simHits.begin(), itEnd = simHits.end(); it != itEnd; ++it, ++globalSimHitIndex) {
+	unsigned int detId_raw = (*it).detUnitId();
+        if (detectorUnits_.find(detId_raw) == detectorUnits_.end()) continue;
+	if (detIds.insert(detId_raw).second) {
+	  // The insert succeeded, so this detector element has not yet been processed.
+	  const std::string algotype = getAlgoType(detId_raw);
+	  const Phase2TrackerGeomDetUnit* phase2det = detectorUnits_[detId_raw];
+	  // access to magnetic field in global coordinates
+	  GlobalVector bfield = pSetup_->inTesla(phase2det->surface().position());
+	  LogDebug("PixelDigitizer") << "B-field(T) at " << phase2det->surface().position() << "(cm): " 
+	         		     << pSetup_->inTesla(phase2det->surface().position());
+	  if (algomap_.find(algotype) != algomap_.end()) 
+	    algomap_[algotype]->accumulateSimHits(it, itEnd, globalSimHitIndex, tofBin, phase2det, bfield);
+	  else
+	    edm::LogInfo("Phase2TrackerDigitizer") << "Unsupported algorithm: " << algotype;
+	}
+      }
+    }
+  }
+  
+  void
+  Phase2TrackerDigitizer::initializeEvent(edm::Event const& e, edm::EventSetup const& iSetup) {
+    
+    // Must initialize all the algorithms
+    for (auto const & el : algomap_) {
+      if (first_) el.second->init(iSetup); 
+      el.second->initializeEvent(); 
+    }
+    first_ = false;
+    // Make sure that the first crossing processed starts indexing the sim hits from zero.
+    // This variable is used so that the sim hits from all crossing frames have sequential
+    // indices used to create the digi-sim link (if configured to do so) rather than starting
+    // from zero for each crossing.
+    crossingSimHitIndexOffset_.clear();
+  
+    iSetup.get<TrackerDigiGeometryRecord>().get(geometryType_, pDD_);
+    iSetup.get<IdealMagneticFieldRecord>().get(pSetup_);
+    iSetup.get<IdealGeometryRecord>().get(tTopoHand);
+    
+    // FIX THIS! We only need to clear and (re)fill this map when the geometry type IOV changes.  Use ESWatcher to determine this.
+    if (true) { // Replace with ESWatcher 
+      detectorUnits_.clear();
+      for (auto const & det_u : pDD_->detUnits()) {
+	unsigned int detId_raw = det_u->geographicalId().rawId();
+	DetId detId = DetId(detId_raw);
+	if (DetId(detId).det() == DetId::Detector::Tracker) {
+	  const Phase2TrackerGeomDetUnit* pixdet = dynamic_cast<const Phase2TrackerGeomDetUnit*>(det_u);
+	  assert(pixdet);
+	  detectorUnits_.insert(std::make_pair(detId_raw, pixdet));
```

Comment:
(as your comment indicates..) the detectorUnits_ should go into a ESSource. But it can at least just be moved to the lumi block instead of event level.